src/main/java/nus/climods/commons/core/CommandSession.java

```diff
@@ -0,0 +1,102 @@
+package nus.climods.commons.core;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+
+import nus.climods.logic.Logic;
+import nus.climods.logic.commands.CommandResult;
+import nus.climods.logic.commands.exceptions.CommandException;
+import nus.climods.logic.parser.exceptions.ParseException;
+
+/**
+ * Keeps the state of commands in application
+ */
+public class CommandSession {
+
+    private final CommandExecutor commandExecutor;
+
+    private final List<String> commandHistory;
+    private ListIterator<String> commandScroller;
+
+    /**
+     * Constructor for CommandSession class.
+     *
+     * @param commandExecutor a function that is responsible for executing commands
+     */
+    public CommandSession(CommandExecutor commandExecutor) {
+        this.commandHistory = new ArrayList<>();
+        this.commandExecutor = commandExecutor;
+    }
+
+    private void resetCommandScroller() {
+        commandScroller = commandHistory.listIterator(commandHistory.size());
+    }
+
+    private void addCommand(String command) {
+        // Only add to command history if command is different from previous command
+        // This is to replicate the behavior on zsh
+        if (commandHistory.isEmpty() || !command.equals(commandHistory.get(commandHistory.size() - 1))) {
+            this.commandHistory.add(command);
+        }
```

Comment:
Note that `bash` does not implement this behavior. In `zsh` if you do not have duplicate consecutive commands in the history, which personally i think it is good since it prevents clutter and bloating

Let me know what you guys think of this behavior :D

Reply:
neat! I am abit conflicted on this: I think this makes a lot more sense, but at the same time most users are probably more used to the other behavior. Still, I think we should go with the no-consecutive-duplicates behaviour, just take note to include this in the UG in case ppl flag this as a bug

Reply:
LGTM 👍  I agree with Cheng Yi too, think that we should state the behaviour in UG jic users flag it as a bug

Reply:
Sounds good! The behavior is now documented in the UG